mobilecoind/src/sync.rs

```diff
@@ -286,8 +296,118 @@ fn sync_thread_entry_point(
+                // We don't expect to get Error::NotFound here, since we earlier tested
+                // ledger_db.num_blocks(). The other error cases are not common
+                // either.
+                let block_idx = monitor_data.next_block + worker_idx as u64;
+                let block_contents = match ledger_db.get_block_contents(block_idx) {
+                    Ok(block_contents) => block_contents,
+                    // Note: mc_ledger_db::Error::NotFound is being handled like any other error
+                    // here, since it really isn't expected.
+                    Err(err) => {
+                        return Err(err.into());
+                    }
+                };
+
+                log::trace!(
+                    logger,
+                    "processing {} outputs and {} key images from block {} for monitor_id {}",
+                    block_contents.outputs.len(),
+                    block_contents.key_images.len(),
+                    block_idx,
+                    monitor_id,
+                );
+
+                // Match tx outs into UTXOs.
+                let utxos = match_tx_outs_into_utxos(
+                    mobilecoind_db,
+                    &block_contents.outputs,
+                    monitor_id,
+                    &monitor_data,
+                    logger,
+                )?;
+
+                Ok((utxos, block_contents.key_images))
+            },
+        )
+        .collect::<Vec<Result<(Vec<UnspentTxOut>, Vec<KeyImage>), Error>>>();
+
+    // For diagnostics, we want to keep track of which workers were successful.
+    // Usually it should be all of them, but if sometimes a worker in the middle
+    // fails and later ones don't, that will harm performance.
+    let worker_successes = parallel_results
+        .iter()
+        .map(|result| if result.is_ok() { 1 } else { 0 })
+        .collect::<Vec<usize>>();
+
+    // Now add everything to the database
+    for (worker_idx, result) in parallel_results.into_iter().enumerate() {
+        match result {
+            Ok((utxos, key_images)) => {
+                // Update database.
+                mobilecoind_db.block_processed(
+                    monitor_id,
+                    monitor_data.next_block + worker_idx as u64,
+                    &utxos,
+                    &key_images,
+                )?;
+            }
+            Err(err) => {
+                // Unfortunately, we have to abandon any work that could have been accomplished
+                // successfully by the threads after this one. To track this,
+                // we'll log a warning about work being abandoned.
+                let abandoned_successes: usize = worker_successes.iter().skip(worker_idx + 1).sum();
+                if abandoned_successes > 0 {
+                    log::warn!(logger, "Due to an error while parallel scanning, had to abandon {} successful block scanning results", abandoned_successes);
```

Comment:
I haven't seen this warning appear yet, but I will test a little more and report back. I don't expect that it should appear.

Reply:
i tried testing against testnet which should be a better test, but running into attestation issues.